File: ru/codes/java/chapter_hashing/simple_hash.java
/**
 * File: simple_hash.java
 * Created Time: 2023-06-21
 * Author: krahets (krahets@163.com)
 */

package chapter_hashing;

public class simple_hash {
    /* Аддитивное хеширование */
    static int addHash(String key) {
        long hash = 0;
        final int MODULUS = 1000000007;
        for (char c : key.toCharArray()) {
            hash = (hash + (int) c) % MODULUS;
        }
        return (int) hash;
    }

    /* Мультипликативное хеширование */
    static int mulHash(String key) {
        long hash = 0;
        final int MODULUS = 1000000007;
        for (char c : key.toCharArray()) {
            hash = (31 * hash + (int) c) % MODULUS;
        }
        return (int) hash;
    }

    /* XOR-хеширование */
    static int xorHash(String key) {
        int hash = 0;
        final int MODULUS = 1000000007;
        for (char c : key.toCharArray()) {
            hash ^= (int) c;
        }
        return hash & MODULUS;
    }

    /* Хеширование с циклическим сдвигом */
    static int rotHash(String key) {
        long hash = 0;
        final int MODULUS = 1000000007;
        for (char c : key.toCharArray()) {
            hash = ((hash << 4) ^ (hash >> 28) ^ (int) c) % MODULUS;
        }
        return (int) hash;
    }

    public static void main(String[] args) {
        String key = "Hello Algo";

        int hash = addHash(key);
        System.out.println("Хеш-сумма сложением = " + hash);

        hash = mulHash(key);
        System.out.println("Хеш-сумма умножением = " + hash);

        hash = xorHash(key);
        System.out.println("Хеш-сумма XOR = " + hash);

        hash = rotHash(key);
        System.out.println("Хеш-сумма с циклическим сдвигом = " + hash);
    }
}

File: ru/codes/kotlin/chapter_hashing/simple_hash.kt
/**
 * File: simple_hash.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package chapter_hashing

/* Аддитивное хеширование */
fun addHash(key: String): Int {
    var hash = 0L
    val MODULUS = 1000000007
    for (c in key.toCharArray()) {
        hash = (hash + c.code) % MODULUS
    }
    return hash.toInt()
}

/* Мультипликативное хеширование */
fun mulHash(key: String): Int {
    var hash = 0L
    val MODULUS = 1000000007
    for (c in key.toCharArray()) {
        hash = (31 * hash + c.code) % MODULUS
    }
    return hash.toInt()
}

/* XOR-хеширование */
fun xorHash(key: String): Int {
    var hash = 0
    val MODULUS = 1000000007
    for (c in key.toCharArray()) {
        hash = hash xor c.code
    }
    return hash and MODULUS
}

/* Вращательное хеширование */
fun rotHash(key: String): Int {
    var hash = 0L
    val MODULUS = 1000000007
    for (c in key.toCharArray()) {
        hash = ((hash shl 4) xor (hash shr 28) xor c.code.toLong()) % MODULUS
    }
    return hash.toInt()
}

/* Driver Code */
fun main() {
    val key = "Hello Algo"

    var hash = addHash(key)
    println("Аддитивное хешированиезначениеравно $hash")

    hash = mulHash(key)
    println("Мультипликативное хешированиезначениеравно $hash")

    hash = xorHash(key)
    println("XOR-хешированиезначениеравно $hash")

    hash = rotHash(key)
    println("Вращательное хешированиезначениеравно $hash")
}
